### ISSUE TITLE
[naga wgsl-out] `let`declarations are wrongly detected as const

### ISSUE THREAD

**Issue by sagudev** [gfx-rs/wgpu#6207]
**Description**
In wgsl-in there is https://github.com/gfx-rs/wgpu/blob/07becfe5b5cc550b04ecbfa303cb6ccfc91357ad/naga/src/front/wgsl/lower/mod.rs#L1370-L1376

but in wgsl-out when we create expression tracker from arena this information is lost:
https://github.com/gfx-rs/wgpu/blob/07becfe5b5cc550b04ecbfa303cb6ccfc91357ad/naga/src/back/wgsl/writer.rs#L169
https://github.com/gfx-rs/wgpu/blob/07becfe5b5cc550b04ecbfa303cb6ccfc91357ad/naga/src/back/wgsl/writer.rs#L197

same problem is also in validator https://github.com/gfx-rs/wgpu/blob/07becfe5b5cc550b04ecbfa303cb6ccfc91357ad/naga/src/valid/function.rs#L1368
~~which causes https://github.com/gfx-rs/wgpu/issues/4390~~ (such checks should only be done in const but let expr should not be treated as const).

**Comment by sagudev** [comment]
> which causes https://github.com/gfx-rs/wgpu/issues/4390

I tested

```patch
diff --git a/naga/src/valid/function.rs b/naga/src/valid/function.rs
index 23e6204cc..c60f3a7e9 100644
--- a/naga/src/valid/function.rs
+++ b/naga/src/valid/function.rs
@@ -1365,7 +1365,11 @@ impl super::Validator {
     ) -> Result<FunctionInfo, WithSpan<FunctionError>> {
         let mut info = mod_info.process_function(fun, module, self.flags, self.capabilities)?;
 
-        let local_expr_kind = crate::proc::ExpressionKindTracker::from_arena(&fun.expressions);
+        let mut local_expr_kind = crate::proc::ExpressionKindTracker::from_arena(&fun.expressions);
+
+        for (handle, _) in fun.expressions.iter() {
+            local_expr_kind.force_non_const(handle);
+        }
 
         for (var_handle, var) in fun.local_variables.iter() {
             self.validate_local_var(var, module.to_ctx(), &info, &local_expr_kind)
```

and it doesn't fix, but we should still need correct constness information for fixing the issue.

**Comment by teoxoy** [comment]
I thought this should be fine for wgsl-out and the validator. The reason it's not ok in the frontend is because treating them as const changes semantics in a non-spec compliant way. What would be an example where this is an issue?

**Comment by sagudev** [comment]
I do not have concrete example but https://github.com/gfx-rs/wgpu/blob/960555a426b0aae48ed36889b8c7feb48de19a6d/naga/src/back/wgsl/writer.rs#L1113-L1125 could upgrade let expressions to const, which might change behavior (for example when overflows are reported, if we did that correctly).

**Comment by teoxoy** [comment]
I see, that would be a problem indeed. I think the wgsl backend should just not emit function-local `const`s at all (that would mean reverting https://github.com/gfx-rs/wgpu/pull/6156/commits/6125ab479056ec2ee859b436a20ed70caac77778 & https://github.com/gfx-rs/wgpu/pull/6156/commits/8e967dbc06cacf9c54969579114f7094cbb71ede).
As for the validator, I think it should be fine as we aren't evaluating those expressions and we already check constness in the frontend.

**Comment by sagudev** [comment]
The problem is that both `let a = 5;` and `const a = 5` will create same result in IR (named expr), but there is no way to now if expr was defined as (local) const (or is just const-able). I think we will need to refactor https://github.com/gfx-rs/wgpu/blob/b7d1f4c6cfbad437c1fd23b3fe53cfb3cc69cbda/naga/src/lib.rs#L1397 to also account for local const or smth like that.

**Comment by sagudev** [comment]
Or alternatively do the reverts and remove const from `named_expressions` as we use local_table for lookup anyway (and let backends inline local const like it's done in https://github.com/gfx-rs/wgpu/pull/7213):
https://github.com/gfx-rs/wgpu/blob/93f64dc847a0e6788b17b292ae4d7d59361e3be0/naga/src/front/wgsl/lower/mod.rs#L1587-L1590

**Comment by teoxoy** [comment]
Not having them in `named_expressions` sounds fine to me.

**Comment by cwfitzgerald** [comment]
Will this have negative consequences on our ability to generate debug information?

**Comment by sagudev** [comment]
> Will this have negative consequences on our ability to generate debug information?

What exactly is considered as debug info in naga?

**Comment by cwfitzgerald** [comment]
I don't think this has a strict definition yet, nor is it set in stone. I think currently the spirv backend uses the variable name in the named_expressions to add label the name of certain values to spirv when debug information is enabled. I would be willing to accept this is the wrong place to do it, and I don't know the specifics, but there are spirv-out tests with debug information in the snapshots.

**Comment by jimblandy** [comment]
Leaving local `const` declarations out of `named_expressions` entirely would be fine with me. When our constant expression and abstract type handling gets better reconciled with validation, we should be able to keep them without any difficulty.
